File: pkg/dart2cpp/sample/dart/hello_restored.dart
import 'package:dart2cpp/platform/dart/runtime_classes.dart';

class GreeterClassInfo extends ClassInfo {
  GreeterClassInfo() {
    greet = Greeter_greet;
    sayHello = Greeter_sayHello;
  }
  String Function(GreeterValue this_)? greet;
  void Function(GreeterValue this_)? sayHello;
}

class GreeterValue extends AnyGC {
  late String name;
  @override
  ClassInfo get classInfo => ClassInfoRegistry.get<GreeterClassInfo>(runtimeType, GreeterClassInfo.new);
}

GreeterValue Greeter_new(AnyGC this__, String name) {
  final this_ = this__ as GreeterValue;
  this_.name = name;
  return this_;
}

String Greeter_greet(AnyGC this__) {
  final this_ = this__ as GreeterValue;
  return 'Hello, ${this_.name}!';
}

void Greeter_sayHello(AnyGC this__) {
  final this_ = this__ as GreeterValue;
  staticPrint((this_.classInfo as GreeterClassInfo).greet!(this_));
}


void main() {
  final GreeterValue greeter = Greeter_new(GC.allocateLocal(GreeterValue()), 'World');
  (greeter.classInfo as GreeterClassInfo).sayHello!(greeter);
  final StaticList<String> names = StaticList<String>.of(['Alice', 'Bob', 'Charlie']);
{
    StaticIterator<String> sync_for_iterator = StaticIterator(names.iterator);
    for (; sync_for_iterator.moveNext(); ) {
      final String name = sync_for_iterator.current;
{
        final GreeterValue g = Greeter_new(GC.allocateLocal(GreeterValue()), name);
        (g.classInfo as GreeterClassInfo).sayHello!(g);
      }
    }
  }
  drainScheduler();
}

File: pkg/dart2cpp/sample/output/hello_restored.cpp
#include "dart2cpp_lowered.h"

struct GreeterValue;
void main();

struct GreeterValue : VPtr {
    std::string name{""};

    void gcMark(int flag) override {
        if (gcFlag == flag) return;
        VPtr::gcMark(flag);
    }
};


GreeterValue* Greeter_new(GreeterValue* this__, std::string name) {
    auto this_ = this__;
    this_->name = name;
    Object_new(this_);
    // unsupported: EmptyStatement
    return this_;
}

std::string Greeter_greet(AnyPtr this__) {
    auto this_ = static_cast<GreeterValue*>(this__.toVPtr());
    return dart_str("Hello, ") + dart_str(this_->name) + dart_str("!");
}

void Greeter_sayHello(AnyPtr this__) {
    auto this_ = static_cast<GreeterValue*>(this__.toVPtr());
    staticPrint((reinterpret_cast<AnyPtr(*)(AnyPtr)>(this_.toVPtr()->vptr["greet"]))(this_));
}

void main() {
    GreeterValue* greeter = Greeter_new(GC::allocateLocal(new GreeterValue()), "World");
    (reinterpret_cast<AnyPtr(*)(AnyPtr)>(greeter.toVPtr()->vptr["sayHello"]))(greeter);
    StaticList<std::string>* names = _literal3("Alice", "Bob", "Charlie");
    StaticIterator<std::string>* sync_for_iterator = names->iterator;
    while ((reinterpret_cast<AnyPtr(*)(AnyPtr)>(sync_for_iterator.toVPtr()->vptr["moveNext"]))(sync_for_iterator)) {
        std::string name = sync_for_iterator->current;
        GreeterValue* g = Greeter_new(GC::allocateLocal(new GreeterValue()), name);
        (reinterpret_cast<AnyPtr(*)(AnyPtr)>(g.toVPtr()->vptr["sayHello"]))(g);
    }
}

